### ISSUE TITLE
Non-determinism is found 

### ISSUE THREAD

**Issue by de-shyt** [JetBrains/lincheck#346]
The code from [here](https://github.com/de-shyt/kotlin-channels/tree/lincheck-non-determinism) produces the following output:

```
WARNING: A Java agent has been loaded dynamically (/Users/Daria.Shutina/.gradle/caches/modules-2/files-2.1/net.bytebuddy/byte-buddy-agent/1.14.12/be4984cb6fd1ef1d11f218a648889dfda44b8a15/byte-buddy-agent-1.14.12.jar)
WARNING: If a serviceability tool is in use, please run with -XX:+EnableDynamicAgentLoading to hide this warning
WARNING: If a serviceability tool is not in use, please run with -Djdk.instrument.traceUsage for more information
WARNING: Dynamic loading of agents will be disallowed by default in a future release

Non-determinism found. Probably caused by non-deterministic code (WeakHashMap, Object.hashCode, etc).
== Reporting the first execution without execution trace ==
= Invalid execution results =
| ------------------------------------------------------------------------------------------------------------------------- |
|                      Thread 1                      |             Thread 2             |             Thread 3              |
| ------------------------------------------------------------------------------------------------------------------------- |
| send(3) + cancel: SUSPENDED + CANCELLED            | receive() + cancel: 3 [3,0,1]    | receive(): SUSPENDED + 2 [2,0,0]  |
| send(3) + cancel: IllegalStateException #1 [1,0,0] | receive(): SUSPENDED + 2 [3,1,1] | send(3): SUSPENDED + void [3,0,1] |
| send(2) + cancel: void [2,0,0]                     | receive(): S [3,2,3]             | send(2): void [3,1,2]             |
| ------------------------------------------------------------------------------------------------------------------------- |

---
Values in "[..]" brackets indicate the number of completed operations
in each of the parallel threads seen at the beginning of the current operation
---
The number next to an exception name helps you find its stack trace provided after the interleaving section
---
Exception stack traces:
#1: java.lang.IllegalStateException: cur.id=0, prev.id=-1, next.id=null
this.id=0
	at com.deshyt.ChannelSegment.findPrev(ChannelSegment.kt:113)
	at com.deshyt.ChannelSegment.removeSegment(ChannelSegment.kt:97)
	at com.deshyt.ChannelSegment.increaseInterruptedCellsCounter$kotlin_channels(ChannelSegment.kt:69)
	at com.deshyt.AtomicWaiter.onInterrupt$kotlin_channels(AtomicWaiter.kt:53)
	at com.deshyt.RendezvousChannel$trySuspendRequest$2$1.invoke(RendezvousChannel.kt:151)
	at com.deshyt.RendezvousChannel$trySuspendRequest$2$1.invoke(RendezvousChannel.kt:151)
	at kotlinx.coroutines.CancelHandler$UserSupplied.invoke(CancellableContinuationImpl.kt:667)
	at kotlinx.coroutines.CancellableContinuationImpl.callCancelHandler(CancellableContinuationImpl.kt:239)
	at kotlinx.coroutines.CancellableContinuationImpl.cancel(CancellableContinuationImpl.kt:209)


== Reporting the second execution ==
= Invalid execution results =
| ----------------------------------------------------------------------------------------------------------------------------- |
|                      Thread 1                      |                     Thread 2                      |       Thread 3       |
| ----------------------------------------------------------------------------------------------------------------------------- |
| send(3) + cancel: SUSPENDED + CANCELLED            | receive() + cancel: SUSPENDED + CANCELLED [3,0,0] | receive(): S [3,0,0] |
| send(3) + cancel: IllegalStateException #1 [1,0,0] | receive(): S [3,1,0]                              | send(3): - [3,2,1]   |
| send(2) + cancel: SUSPENDED + CANCELLED [2,0,0]    | receive(): - [3,2,0]                              | send(2): - [3,2,1]   |
| ----------------------------------------------------------------------------------------------------------------------------- |

```

The whole output can be found in the [log file](https://github.com/de-shyt/kotlin-channels/blob/lincheck-non-determinism/lincheck-non-determinism-output.log) in the same branch

**Comment by ndkoval** [comment]
@de-shyt could you please check whether it works with Lincheck 2.33?

**Comment by ndkoval** [comment]
@de-shyt should I close the issue?
